src/ptc/start.clj

```diff
@@ -81,35 +81,38 @@
       (let [peeked (jms/ednify peeked)]
         (do
           (log/infof "Peeked message %s: %s" counter peeked)
-          (if (task! peeked)
+          (when (task! peeked connection)
             (let [consumed (jms/ednify (consume connection queue))]
               (log/infof "Task complete, consumed message %s" counter)
               (if (not (misc/message-ids-equal? peeked consumed))
                 (log/warnf
                  (str/join \space ["Messages differ:"
                                    (with-out-str (pprint (data/diff peeked consumed)))])))
-              (recur (inc counter)))
-            (do
-              (log/errorf
-               (str/join
-                \space ["Task returned nil/false,"
-                        "not consuming message %s and instead exiting"])
-               counter)
-              peeked))))
+              (recur (inc counter))))))
       (recur counter))))
 
 (defn- message-loop
   "Loop and consume messages using the Zamboni ActiveMQ server."
   []
-  (let [queue          (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_QUEUE_NAME")
-        url            (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_SERVER_URL")
-        vault-path     (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_SECRET_PATH")
-        bucket-url     (misc/getenv-or-throw "PTC_BUCKET_URL")
-        upload-sample! (partial jms/handle-message bucket-url)
+  (let [queue               (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_QUEUE_NAME")
+        dead-letter-queue   (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_DEAD_LETTER_QUEUE_NAME")
+        url                 (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_SERVER_URL")
+        vault-path          (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_SECRET_PATH")
+        bucket-url          (misc/getenv-or-throw "PTC_BUCKET_URL")
+        upload-sample-or-move-to-dlq! (fn [jms connection]
+                                        (try
+                                          (jms/handle-message bucket-url jms)
+                                          (catch Throwable x
+                                            (produce connection dead-letter-queue (str misc/utc-now) jms)
+                                            (log/errorf
+                                             (str/join
+                                              \space ["Task returned nil/false,"
+                                                      "not consuming message %s, moving it to dead letter queue and continue..."])))
+                                          (finally true)))
```

Comment:
Log the exception and message.
The ActiveMQ server will timestamp the message, but cannot record the exception, so pass that.
I don't see where PtC consumes the message in the DLQ path.

Suggest:
```clojure
(defn- message-loop
  "Loop and consume messages using the Zamboni ActiveMQ server."
  []
  (let [queue      (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_QUEUE_NAME")
        dlq        (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_DEAD_LETTER_QUEUE_NAME")
        url        (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_SERVER_URL")
        vault-path (misc/getenv-or-throw "ZAMBONI_ACTIVEMQ_SECRET_PATH")
        bucket-url (misc/getenv-or-throw "PTC_BUCKET_URL")
        {:keys [username password]} (misc/vault-secrets vault-path)]
    (letfn [(handle-or-dlq! [jms connection]
              (try (jms/handle-message bucket-url jms)
                   (catch Throwable x
                     (log/errorf x "Moving to %s: %s" dlq jms)
                     (produce connection dlq (str x) jms))
                   (finally true)))]
      (try
        (with-open [connection (create-queue-connection url username password)]
          (listen-and-consume-from-queue handle-or-dlq! connection queue))
        (catch Throwable x
          (log/fatal x "Fatal error in message loop")
          (System/exit 1))))))
```

Reply:
I see where PtC consumes the message in the DLQ path now.